Add base classes for output plugins.

dmikurube opened the pull request.

dmikurube commented:
@muga @sakama Sorry that it's a big change -- but I couldn't split since they're basically all linked.

Can you have a look? We can have a walk-through together if needed.

sakama commented on embulk-output-example/src/main/java/org/embulk/output/example/ExampleOutputPluginDelegate.java:
```diff
@@ -0,0 +1,125 @@
+    {
+    }
+
+    // @Override  // Overridden from |ServiceRequestMapperBuildable|
```
Remove comment out?

dmikurube replied:
Ah, exactly! Will fix.

dmikurube replied:
Fixed.

sakama commented on embulk-output-example/src/main/java/org/embulk/output/example/ExampleOutputPluginDelegate.java:
```diff
@@ -0,0 +1,125 @@
+
+        ObjectNode json = JsonNodeFactory.instance.objectNode();
+        json.put("timestamp", Calendar.getInstance().getTime().getTime());
+        json.set("records", records);
```
`put(String fieldName, JsonNode value)` was deprecated.
```
* @deprecated Since 2.4 use either {@link #set(String,JsonNode)} or {@link #replace(String,JsonNode)},*/
@Deprecated
public JsonNode put(String fieldName, JsonNode value)
{
    ...
}
```
Use `replace(String fieldName, JsonNode value)` is better?

dmikurube replied:
Ah, yes, it's now `set`.

dmikurube commented:
Added some comments and Javadocs.

dmikurube commented:
Updated: `TaskValidatable` supported only validating a `Task` for input plugins. Output plugins may need to validate `Schema` and `taskCount` as well.

Added new `OutputTaskValidatable` for output plugins.

`TaskValidatable` will be renamed to `InputTaskValidatable`.

muga commented:
LGTM 👍  

The library supports the development of output plugins that execute single output task and egest data after Embulk transaction. It's good because almost all of plugins might be covered. If we could in the future, it's even better to support output plugins that execute multiple tasks and upload data on the tasks in parallel like embulk-output-elasticsearch.

dmikurube commented:
Thanks! Will be merging.

Designed the basic interfaces so that it can support parallel uploads, but we may need to confirm what is required to implement parallel uploads by multiple tasks.
